### PR TITLE
Noise augmentation tutorial

### PR DESCRIPTION
**Pull request by jbalam-nv**

### CODE REVIEW

File: asr-python-advanced-finetune-am-noise-augmentation.ipynb [hunk 1]
```diff
@@ -0,0 +1,541 @@
...
+    "##### We don't need the NeMo container anymore, so let's get rid of it"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "###  Fine-tuning the ASR model\n",
+    "To finetune the ASR model with the augmented datasets that we just created, you can \n",
+    "proceed to [this tutorial](https://github.com/nvidia-riva/tutorials/blob/stable/asr-python-advanced-finetune-am-conformer-ctc-tao-finetuning.ipynb)."
```
Comment (jbalam-nv) [inline asr-python-advanced-finetune-am-noise-augmentation.ipynb:500]: This path needs to be changed to the path of the other finetuning tutorial after it is merged.

File: asr-python-advanced-finetune-am-noise-augmentation.ipynb [hunk 2]
```diff
@@ -0,0 +1,541 @@
...
+    "tags": []
+   },
+   "source": [
+    "#### Download and process the AN4 dataset\n",
+    "AN4 is a small dataset recorded and distributed by Carnegie Mellon University (CMU). It consists of recordings of people spelling out addresses, names, etc. Information about this dataset can be found on the official CMU site.\n",
+    "\n",
+    "Let's download the AN4 dataset tar file."
+   ]
+  },
+  {
```
Comment (svenchilton) [inline asr-python-advanced-finetune-am-noise-augmentation.ipynb:80]:
```suggestion
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Install the necessary dependencies\n",
    "!pip install wget\n",
    "!apt-get install sox libsndfile1 ffmpeg libsox-fmt-mp3\n",
    "!pip install text-unidecode\n",
    "!pip install matplotlib>=3.3.2\n",
    "!pip install Cython\n",
    "\n",
    "# Import the necessary dependencies.\n",
    "import wget\n",
    "import glob\n",
    "import os\n",
    "import subprocess\n",
    "import tarfile"
   ]
  },
  {
```

**Review by svenchilton — APPROVED** [review]
Split the first code cell, added pip install statements to install the necessary dependencies, changed all instances of "Please proceed" to "Proceed"

File: asr-noise-augmentation.ipynb [hunk 3]
```diff
@@ -0,0 +1,605 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<img src=\"http://developer.download.nvidia.com/notebooks/dlsw-notebooks/rivaasrasr-noise-augmentation/nvidia_logo.png\" style=\"width: 90px; float: right;\">\n",
+    "\n",
+    "# How to improve accuracy on noisy speech by fine-tuning the Acoustic Model (Conformer-CTC) in the Riva ASR pipeline \n",
```
Comment (LynseyFabel) [inline asr-noise-augmentation.ipynb:9]:
```suggestion
    "# How to Improve the Accuracy on Noisy Speech by Fine-Tuning the Acoustic Model (Conformer-CTC) in the Riva ASR Pipeline \n",
```

File: asr-noise-augmentation.ipynb [hunk 4]
```diff
@@ -0,0 +1,605 @@
...
+    "\n",
+    "This tutorial walks you through some of the advanced customization features of the Riva ASR pipeline by fine-tuning the Acoustic Model (Conformer-CTC). These customization features improve accuracy on specific speech scenarios, like background noise and different acoustic environments."
```
Comment (LynseyFabel) [inline asr-noise-augmentation.ipynb:11]:
```suggestion
    "This tutorial walks you through some of the advanced customization features of the Riva ASR pipeline by fine-tuning the acoustic model (Conformer-CTC). These customization features improve accuracy on specific speech scenarios, like background noise and different acoustic environments."
```

File: asr-noise-augmentation.ipynb [hunk 5]
```diff
@@ -0,0 +1,605 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## NVIDIA Riva Overview\n",
+    "\n",
+    "NVIDIA Riva is a GPU-accelerated SDK for building Speech AI applications that are customized for your use case and deliver real-time performance. <br/>\n",
```
Comment (LynseyFabel) [inline asr-noise-augmentation.ipynb:20]:
```suggestion
    "NVIDIA Riva is a GPU-accelerated SDK for building speech AI applications that are customized for your use case and deliver real-time performance. <br/>\n",
```

File: asr-noise-augmentation.ipynb [hunk 6]
```diff
@@ -0,0 +1,605 @@
...
+    "Riva offers a rich set of speech and natural language understanding services such as:\n",
+    "\n",
+    "- Automated speech recognition (ASR)\n",
+    "- Text-to-Speech synthesis (TTS)\n",
+    "- A collection of natural language processing (NLP) services, such as named entity recognition (NER), punctuation, and intent classification.\n",
+    "\n",
+    "In this tutorial, we will show how to augment your training data with background noise data for fine-tuning the Acoustic Model (Conformer-CTC) to improve accuracy on audio with background noise.  \n",
```
Comment (LynseyFabel) [inline asr-noise-augmentation.ipynb:27]:
```suggestion
    "In this tutorial, we will show how to augment your training data (with background noise data) for fine-tuning the acoustic model (Conformer-CTC) to improve accuracy on audio with background noise.  \n",
```

File: asr-noise-augmentation.ipynb [hunk 7]
```diff
@@ -0,0 +1,605 @@
...
+    "To understand the basics of Riva ASR APIs, refer to [Getting started with Riva ASR in Python](https://github.com/nvidia-riva/tutorials/blob/stable/asr-python-basics.ipynb). <br>\n",
+    "\n",
+    "For more information about Riva, refer to the Riva [product page](https://developer.nvidia.com/riva) and [documentation](https://docs.nvidia.com/deeplearning/riva/user-guide/docs/overview.html)."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Data preprocessing\n",
```
Comment (LynseyFabel) [inline asr-noise-augmentation.ipynb:37]:
```suggestion
    "### Data Preprocessing\n",
```

File: asr-noise-augmentation.ipynb [hunk 8]
```diff
@@ -0,0 +1,605 @@
...
+    "\n",
+    "For fine-tuning we need audio data with background noise. If you already have such data, then you can use it directly.  \n",
```
Comment (LynseyFabel) [inline asr-noise-augmentation.ipynb:39]:
```suggestion
    "For fine-tuning, we need audio data with background noise. If you already have such data, then you can use it directly.  \n",
```

File: asr-noise-augmentation.ipynb [hunk 9]
```diff
@@ -0,0 +1,605 @@
...
+    "In this tutorial, we will take the AN4 dataset and augment it with noise data from the Room Impulse Response and Noise Database from the [openslr database](https://www.openslr.org/28/).\n",
```
Comment (LynseyFabel) [inline asr-noise-augmentation.ipynb:40]:
```suggestion
    "In this tutorial, we will take the AN4 dataset and augment it with noise data from the Room Impulse Response and Noise Database from the [OpenSLR database](https://www.openslr.org/28/).\n",
```

File: asr-noise-augmentation.ipynb [hunk 10]
```diff
@@ -0,0 +1,605 @@
...
+    "\n",
+    "In this tutorial, we will be using NVIDIA NeMo for the data preprocessing step.\n",
+    "\n",
+    "#### NVIDIA NeMo Overview\n",
+    "\n",
+    "NVIDIA NeMo is a toolkit for building new state-of-the-art conversational AI models. NeMo has separate collections for Automatic Speech Recognition (ASR), Natural Language Processing (NLP), and Text-to-Speech (TTS) models. Each collection consists of prebuilt modules that include everything needed to train on your data. Every module can easily be customized, extended, and composed to create new conversational AI model architectures.\n",
```
Comment (LynseyFabel) [inline asr-noise-augmentation.ipynb:46]: "...train on your data" seems odd to me. You train your model with your data.  What are your training on your data?

File: asr-noise-augmentation.ipynb [hunk 11]
```diff
@@ -0,0 +1,605 @@
...
+    "For more information about NeMo, refer to the [NeMo product page](https://developer.nvidia.com/nvidia-nemo) and [documentation](https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/stable/starthere/intro.html). The open-source NeMo repository can be found [here](https://github.com/NVIDIA/NeMo).\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Requirements and setup for data preprocessing:"
```
Comment (LynseyFabel) [inline asr-noise-augmentation.ipynb:54]:
```suggestion
    "#### Requirements and Setup for Data Preprocessing:"
```

File: asr-noise-augmentation.ipynb [hunk 12]
```diff
@@ -0,0 +1,605 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "##### Requirements:\n",
+    "\n",
+    "We will be using [NVIDIA NeMo](https://github.com/NVIDIA/NeMo) for this data preprocessing step - Easiest way to install and run NeMo is through NVIDIA's PyTorch docker container. If you are not already running this notebook in the NVIDIA PyTorch docker container, please follow instructions [here](./README.md#running-the-nvidia-riva-tutorial-how-to-improve-accuracy-on-specific-speech-patterns-by-fine-tuning-the-acoustic-model-citrinet-in-the-riva-asr-pipeline) to re-run this tutorial from the PyTorch docker container\n",
```
Comment (LynseyFabel) [inline asr-noise-augmentation.ipynb:63]:
```suggestion
    "We will be using [NVIDIA NeMo](https://github.com/NVIDIA/NeMo) for this data preprocessing step. The easiest way to install and run NeMo is through NVIDIA's PyTorch Docker container. If you are not already running this tutorial in the NVIDIA PyTorch Docker container, follow the instructions to do [here](./README.md#running-the-nvidia-riva-tutorial-how-to-improve-accuracy-on-specific-speech-patterns-by-fine-tuning-the-acoustic-model-citrinet-in-the-riva-asr-pipeline).\n",
```

File: asr-noise-augmentation.ipynb [hunk 13]
```diff
@@ -0,0 +1,605 @@
...
+    "\n",
+    "We will be using the NVIDIA NeMo Docker container, so access to NGC is a must. "
```
Comment (LynseyFabel) [inline asr-noise-augmentation.ipynb:65]:
```suggestion
    "We will be using the NVIDIA NeMo Docker container, therefore, its required that you have access to NGC. "
```

File: asr-noise-augmentation.ipynb [hunk 14]
```diff
@@ -0,0 +1,605 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "#### Download and process the AN4 dataset\n",
```
Comment (LynseyFabel) [inline asr-noise-augmentation.ipynb:74]:
```suggestion
    "#### Download and Process the AN4 Dataset\n",
```

File: asr-noise-augmentation.ipynb [hunk 15]
```diff
@@ -0,0 +1,605 @@
...
+    "    print(\"Test manifest already exists at\", test_manifest)\n",
+    "\n",
+    "print(\"***Done***\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Download and process the background noise dataset\n",
```
Comment (LynseyFabel) [inline asr-noise-augmentation.ipynb:244]:
```suggestion
    "#### Download and Process the Background Noise Dataset\n",
```

File: asr-noise-augmentation.ipynb [hunk 16]
```diff
@@ -0,0 +1,605 @@
...
+    "\n",
+    "For background noise, we will use the background noise samples from the Room Impulse Response and Noise database from the openslr database. For each 30 second isotropic noise sample in the dataset we use the first 15 seconds for training and the last 15 seconds for evaluation.\n",
```
Comment (LynseyFabel) [inline asr-noise-augmentation.ipynb:246]:
```suggestion
    "For background noise, we will use the background noise samples from the Room Impulse Response and Noise database from the OpenSLR database. For each 30 second isotropic noise sample in the dataset, we use the first 15 seconds for training and the last 15 seconds for evaluation.\n",
```

File: asr-noise-augmentation.ipynb [hunk 17]
```diff
@@ -0,0 +1,605 @@
...
+    "\n",
+    "Let's first download this dataset."
```
Comment (LynseyFabel) [inline asr-noise-augmentation.ipynb:248]:
```suggestion
    "Let's first download the dataset."
```

File: asr-noise-augmentation.ipynb [hunk 18]
```diff
@@ -0,0 +1,605 @@
...
+    "    write_manifest(train_noise_manifest, train_rows)\n",
+    "else:\n",
+    "    print('Train noise manifest file already exists. Proceed to the next step.')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Create the noise-augmented dataset\n",
```
Comment (LynseyFabel) [inline asr-noise-augmentation.ipynb:395]:
```suggestion
    "#### Create the Noise-Augmented Dataset\n",
```

File: asr-noise-augmentation.ipynb [hunk 19]
```diff
@@ -0,0 +1,605 @@
...
+    "\n",
+    "Finally, let's create a noise-augmented dataset by adding noise to the the AN4 dataset with the `add_noise.py` NeMo script. This script generates the noise-augmented audio clips as well as the manifest files. \n",
```
Comment (LynseyFabel) [inline asr-noise-augmentation.ipynb:397]:
```suggestion
    "Finally, let's create a noise-augmented dataset by adding noise to the AN4 dataset with the `add_noise.py` NeMo script. This script generates the noise-augmented audio clips as well as the manifest files. \n",
```

File: asr-noise-augmentation.ipynb [hunk 20]
```diff
@@ -0,0 +1,605 @@
...
+    "# Clone NeMo locally\n",
+    "nemo_dir = os.path.join(os.getcwd(), 'NeMo')\n",
+    "!git clone https://github.com/NVIDIA/NeMo $nemo_dir"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "##### Training dataset\n",
```
Comment (LynseyFabel) [inline asr-noise-augmentation.ipynb:434]:
```suggestion
    "##### Training Dataset\n",
```

File: asr-noise-augmentation.ipynb [hunk 21]
```diff
@@ -0,0 +1,605 @@
...
+    "    --noise_manifest=$train_noise_manifest \\\n",
+    "    --snrs 0 5 10 15 \\\n",
+    "    --out_dir=$final_data_dir"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The above script generates a `.json` manifest file each for every SNR value, i.e., one manifest file each for 0, 5, 10 and 15db SNR. \n",
```
Comment (LynseyFabel) [inline asr-noise-augmentation.ipynb:463]:
```suggestion
    "The above script generates a `.json` manifest file each for every SNR value, that is, one manifest file each for 0, 5, 10, and 15db SNR. \n",
```

File: asr-noise-augmentation.ipynb [hunk 22]
```diff
@@ -0,0 +1,605 @@
...
+    "print(\"Combined manifest for noise-augmented training dataset created at\", final_data_dir + \"/manifests/noisy_train_manifest.json\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "##### Test dataset\n",
+    "\n",
+    "Let's create a noise-augmented evaluation dataset using the an4 test dataset, by adding noise at 5 dB, using the same NeMo `add_noise.py` script with which we augmented the training dataset."
```
Comment (LynseyFabel) [inline asr-noise-augmentation.ipynb:506]: Its an4 and AN4 throughout this tutorial. Can we be consistent with just one naming style?

File: asr-noise-augmentation.ipynb [hunk 23]
```diff
@@ -0,0 +1,605 @@
...
+    "**Noise-augmented testing manifest and data are created at `{working_dir}/noise_data/manifests/noisy_test_manifest_5db.json` and `{working_dir}/noise_data/test_manifest` respectively.**  "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "###  Fine-tuning the ASR model\n",
```
Comment (LynseyFabel) [inline asr-noise-augmentation.ipynb:562]:
```suggestion
    "###  Fine-Tuning the ASR Model\n",
```

File: asr-noise-augmentation.ipynb [hunk 24]
```diff
@@ -0,0 +1,605 @@
...
+    "To fine-tune the ASR model with the augmented datasets that we just created, you can \n",
+    "proceed to [this tutorial](https://github.com/nvidia-riva/tutorials/blob/main/asr-finetuning-conformer-ctc-nemo.ipynb). In that case, make sure to reset the manifest and dataset filepaths appropriately when calling the NeMo tokenization, training, and evaluation scripts. "
```
Comment (LynseyFabel) [inline asr-noise-augmentation.ipynb:564]:
```suggestion
    "proceed to [this tutorial](https://github.com/nvidia-riva/tutorials/blob/main/asr-finetuning-conformer-ctc-nemo.ipynb). In this case, make sure to reset the manifest and dataset file paths appropriately when calling the NeMo tokenization, training, and evaluation scripts. "
```

**Review by LynseyFabel — APPROVED** [review]
I've left a few comments and have finished my review. Thank you!!

File: asr-noise-augmentation.ipynb (hunk 22 above)
Comment (svenchilton) [reply asr-noise-augmentation.ipynb:506]: I made a few changes, which I'll commit and push momentarily. It should now be consistently rendered "AN4" in Markdown cells and printouts and "an4" in variable names and file paths.

File: asr-noise-augmentation.ipynb (hunk 12 above)
Comment (svenchilton) [reply asr-noise-augmentation.ipynb]: I rewrote this section. See https://github.com/nvidia-riva/tutorials/pull/120/commits/ddb2fd9b19215e8ec7b2d273e5d7525eee3e6b1b.

File: asr-noise-augmentation.ipynb (hunk 10 above)
Comment (svenchilton) [reply asr-noise-augmentation.ipynb]: I changed "train on your data" to "train conversational AI models on your data". See https://github.com/nvidia-riva/tutorials/pull/120/commits/2d4d8b85ef588889e1238241b0f93d0c3e9ab996.

**Pull request merged**